pkg/autoupdate/findnew.go

```diff
@@ -0,0 +1,177 @@
+package autoupdate
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"path/filepath"
+	"sort"
+	"strings"
+
+	"github.com/go-kit/kit/log"
+	"github.com/go-kit/kit/log/level"
+	"github.com/kolide/launcher/pkg/contexts/ctxlog"
+	"github.com/pkg/errors"
+)
+
+// This suffix is added to the binary path to find the updates
+const updateDirSuffix = "-updates"
+
+type newestSettings struct {
+	deleteOld bool
+}
+
+type newestOption func(*newestSettings)
+
+func DeleteOldUpdates() newestOption {
+	return func(no *newestSettings) {
+		no.deleteOld = true
+	}
+}
+
+// FindNewestSelf invokes `FindNewest` with the running binary path,
+// as determined by os.Executable. However, if the current running
+// version is the same as the newest on disk, it will return empty string.
+func FindNewestSelf(ctx context.Context, opts ...newestOption) (string, error) {
+	logger := log.With(ctxlog.FromContext(ctx), "caller", log.DefaultCaller)
+
+	exPath, err := os.Executable()
+	if err != nil {
+		return "", errors.Wrap(err, "determine running executable path")
+	}
+
+	if exPath == "" {
+		return "", errors.New("can't find newest empty string")
+	}
+
+	newest := FindNewest(ctx, exPath, opts...)
+
+	if newest == "" {
+		return "", nil
+	}
+
+	if exPath == newest {
+		return "", nil
+	}
+
+	level.Debug(logger).Log(
+		"msg", "found an update",
+		"newest", newest,
+		"exPath", exPath,
+	)
+
+	return newest, nil
+}
+
+// FindNewest takes the full path to a binary, and returns the newest
+// update on disk. If there are no updates on disk, it returns the
+// original path. It will return the same fullBinaryPath if that is
+// the newest version.
+func FindNewest(ctx context.Context, fullBinaryPath string, opts ...newestOption) string {
+	logger := log.With(ctxlog.FromContext(ctx), "caller", log.DefaultCaller)
+
+	if fullBinaryPath == "" {
+		level.Debug(logger).Log("msg", "called with empty string")
+		return ""
+	}
+
+	logger = log.With(logger, "fullBinaryPath", fullBinaryPath)
+
+	newestSettings := &newestSettings{}
+	for _, opt := range opts {
+		opt(newestSettings)
+	}
+
+	updateDir := getUpdateDir(fullBinaryPath)
+	binaryName := filepath.Base(fullBinaryPath)
+
+	fileGlob := filepath.Join(updateDir, "*", binaryName)
+
+	possibleUpdates, err := filepath.Glob(fileGlob)
+	if err != nil {
+		level.Error(logger).Log("msg", "globbing for updates", "err", err)
+		return fullBinaryPath
+	}
+
+	sort.Strings(possibleUpdates)
+
+	// iterate backwards over files, looking for a suitable binary
+	foundCount := 0
+	foundFile := ""
+	for i := len(possibleUpdates) - 1; i >= 0; i-- {
+		file := possibleUpdates[i]
+
+		// If we've already found at least 2 files, (newest, and presumed
+		// current), trigger delete routine
+		if newestSettings.deleteOld && foundCount >= 2 {
```

Comment:
I assume that in empirical testing this works as expected, but do we have any guarantees about how the matches from `Glob` are sorted across platforms?

Reply:
it could be that I just need to read the test files still..

Reply:
sort.strings is a couple lines up. https://github.com/kolide/launcher/pull/524/files/df0e753ec23419adb1d0cab980735e038146d562#diff-130b1554d3f180a250c35f46b1473c8fR96

Reply:
ha such a 🤦‍♂  - I was _looking_ for that, but missed it.